plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java

```diff
@@ -40,22 +42,46 @@ public PipelineExecutionService(PipelineStore store, ThreadPool threadPool) {
         this.threadPool = threadPool;
     }
 
-    public void execute(IngestDocument ingestDocument, String pipelineId, Listener listener) {
+    public void execute(IndexRequest indexRequest, String pipelineId, Listener listener) {
         Pipeline pipeline = store.get(pipelineId);
         if (pipeline == null) {
             listener.failed(new IllegalArgumentException("pipeline with id [" + pipelineId + "] does not exist"));
             return;
         }
 
-        threadPool.executor(THREAD_POOL_NAME).execute(new Runnable() {
-            @Override
-            public void run() {
-                try {
-                    pipeline.execute(ingestDocument);
-                    listener.executed(ingestDocument);
-                } catch (Throwable e) {
-                    listener.failed(e);
+        threadPool.executor(THREAD_POOL_NAME).execute(() -> {
+            String index = indexRequest.index();
+            String type = indexRequest.type();
+            String id = indexRequest.id();
+            String routing = indexRequest.routing();
+            String parent = indexRequest.parent();
+            String timestamp = indexRequest.timestamp();
+            String ttl = null;
+            if (indexRequest.ttl() != -1) {
+                ttl = String.valueOf(indexRequest.ttl());
+            }
+            Map<String, Object> sourceAsMap = indexRequest.sourceAsMap();
+            IngestDocument ingestDocument = new IngestDocument(index, type, id, routing, parent, timestamp, ttl, sourceAsMap);
+            try {
+                pipeline.execute(ingestDocument);
+                if (ingestDocument.isSourceModified()) {
+                    indexRequest.source(ingestDocument.getSource());
+                }
+                if (ingestDocument.isMetaDataModified()) {
```

Comment:
I think see the point of not setting the source if it was not modified, but when it comes to metadata, should we just set them back all the time? anyway we end up with a single flag for all of them...